fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,239 @@
+    /// Wait for next change in this source.
+    async fn wait_for_update(&self);
+    async fn fetch(
+        &self,
+        client: &Client,
+        fetch_kind: FetchKind,
+    ) -> anyhow::Result<BTreeMap<MetaFieldKey, MetaFieldValue>>;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FetchKind {
+    /// Meta source should return fast, retry less.
+    /// This blocks getting any meta values.
+    Initial,
+    /// Meta source can retry infinitely.
+    Background,
+}
+
+/// Service for managing the caching of meta fields.
+// a fancy DST to save one allocation.
+pub struct MetaService<S: ?Sized = dyn MetaSource> {
+    initial_fetch_waiter: Waiter,
+    source: S,
+}
+
+impl<S: MetaSource + ?Sized> MetaService<S> {
+    pub fn new(source: S) -> Arc<MetaService>
+    where
+        S: Sized,
+    {
+        // implicit cast `Arc<MetaService<S>>` to `Arc<MetaService<dyn MetaSource>>`
+        Arc::new(MetaService {
+            initial_fetch_waiter: Waiter::new(),
+            source,
+        })
+    }
+
+    /// Get the value for the meta field.
+    ///
+    /// This may wait for significant time on first run.
+    pub async fn get_field<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<V> {
+        if let Some(value) = self.get_field_from_db(db, field).await {
+            // might be from in old cache.
+            // TODO: maybe old cache should have a ttl?
+            Some(value)
+        } else {
+            // wait for initial value
+            self.initial_fetch_waiter.wait().await;
+            self.get_field_from_db(db, field).await
+        }
+    }
+
+    async fn get_field_from_db<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<V> {
+        db.begin_transaction_nc()
+            .await
+            .get_value(&MetaFieldKey(field.to_string()))
+            .await
+            .and_then(|value| parse_meta_value_static::<V>(&value.0).ok())
+    }
+
+    /// Update all source in background.
+    ///
+    /// Caller should run this method in a task.
+    pub(crate) async fn update_continuously(&self, client: &Client) -> ! {
+        let meta = self.source.fetch(client, FetchKind::Initial).await;
+        let failed_initial = meta.is_err();
+        match meta {
+            Ok(meta) => Self::save_meta_values(client, &meta).await,
+            Err(error) => warn!(%error, "failed to fetch source"),
+        };
+        self.initial_fetch_waiter.done();
+
+        // don't wait if we failed first item
+        if !failed_initial {
+            self.source.wait_for_update().await;
+        }
+
+        // now keep updating slowly
+        loop {
+            if let Ok(meta) = self.source.fetch(client, FetchKind::Background).await {
+                Self::save_meta_values(client, &meta).await;
+            }
+            self.source.wait_for_update().await;
+        }
+    }
```

Comment:
Isn't there still the problem that `failed_initial` never gets reset and a `fetch` impl that doesn't have some exp backoff internally will just run this loop at max speed?